### ISSUE TITLE
Update 2017-10-16 ~ Build Well R11 ~ better geometry handling

### ISSUE THREAD

**Issue by theo-armour** [ladybug-tools/spider#45]
@ladybug-tools/spiders 

![image](https://user-images.githubusercontent.com/547626/31642044-9220c1d4-b29d-11e7-9821-7a7c56410776.png)
_Updates to Site Settings menu_

## Full Screen: [Build Well R11]( http://ladybug-tools.github.io/spider/build-well/index.html )

Another big update. Last time it was the menus, this time it was the geometry handling.

The whole package now has quite a few loose ends. The next few pint releases should begin to clean things up a bit

Coming up next is

* Bring the L Shape up to the level of the Box Shape
* Work on Zones and Spaces menus
* Work on Anton's WWR menu
* Separate ft/in and metric menus. Also multiple languages maybe
* gbXML export needs rotation and overhangs 

## Change Log
* R11
* New geometry handling process
* 'The Building' is always in shape 
* Kicks off as a Box Shape from get go
	* Created as a JavaScript object that exists throughout session
	* Geometry can come and go, but defaults and most variables persist
	* Should make it easy to switch between shapes
* Site Settings
	* Now includes Location and Units details
	* Units not yet wired up to other menu

To do list items mostly dealt with
* 2017-10-14 ~ Open with Box Shape already in place
* 2017-09-22 ~ Add lat & Lon

**Comment by theo-armour** [comment]
Build Well R11.1: L Shape beginning to work. Final version awaiting full spec.

**Comment by theo-armour** [comment]
BW R11.2

* mnu-shape-ell-one.html
	* Working somewhat
	* Final version awaiting a full spec
* mnu-gbxml-export.html
	* Better text display
	* Variables now connected to mnu-settings-site.html
* mnu-gbxml-export.html
	* Many fixes, works with new geometry format
	* 2017-10-17 ~ Don't export ground face of adjacent buildings
	* Started process to export overhangs. 
	* @@Note gbXML files not loading into viewer because of as yet incomplete data

**Comment by theo-armour** [comment]
BW R11.3

![image](https://user-images.githubusercontent.com/547626/31746796-06d5331c-b41e-11e7-92d6-acfda916498d.png)

Still many issues

Plus still need to export manifold space (floor + ceiling ) and still can't rotate the building and export.

But we are moving along

**Comment by bwelle** [comment]
@theo-armour 
Great step forward! One thing to clarify...When I say ceiling height I mean floor height. We don't use suspended ceilings in pre-design. That would require another space for the plenum and way more simulation time. It's fine the way you have it. I just meant that apparently in the gbXML file, you call each interior floor surface "Ceiling" only once and put two adjacent spaces to it, above and below. It used to be that you needed to define a floor surface twice, once as a "floor" for a space and once as a "ceiling" for a space, with coordinates reversed. They've simplified now. I don't think there's a relation between which adjacent space you list first or second based on the coordinate ordering. But I will check just in case.

**Comment by bwelle** [comment]
@theo-armour 
Going to start listing some functionality required for the demo, and then put other non-critical path items on a wishlist.

**Comment by bwelle** [comment]
@theo-armour 
First I'll just jump into geometry.  You can ignore that last spreadsheet I posted. We have fundamentally changed the method for geometry to be more robust and flexible for architects to play with shapes. You can find the new logic in an Excel spreadsheet [here](https://perkinswill.box.com/s/0il7fbx5fbemn7lu15rru6oed1ejf881), with some comments to help clarify, but not sure if it's clear so happy to hop on a call to walk you through it.

The spreadsheet has the equations for a Box, L-shape, and H-shape. If we have the example for the L-shape, we can do the H-shape ourselves, but feel free to add it if you want others to use. Critical path is Box and L-shape. 

When you land on the Box page, please put these defaults in this order...Building Area=5000 ft2, Number of Floors=1, Floor Area=calculated, Length uses the equation in the spreadsheet (sqrt(FloorArea)), Width=FloorArea\Length, Volume=calculated, Floor Height=12ft, Perimeter Depth=10 ft.

Make Building Area a slider with a range from 1000-100,000 ft2. Increments of 1000 ft. Make Number of Floors a slider from 1-10, increments of 1. Make Length a slider with the ranges calculated in the spreadsheet (min=sqrt(FloorArea/10), max=sqrt(FloorArea*10). Make Floor Height a slider between 8-20ft, increments of 1 foot. Make Perimeter Depth a slider between 10-25 feet, increments of 1. All other parameters are calculated fields, no need for a slider.

Then have a heading for South, West, East, North. Under each heading put WWR=0.4, # of Windows=4, Window Ratio=0.5, Overhang Depth=3ft, and Fin Depth=0ft.

WWR is a slider between 0-0.9, increments of 0.1. # of Windows a slider between 1-20, increments of 1. Window Ratio a slider between 0.1 and 0.9, increments of 0.1. Overhang Depth a slider between 0-5 ft, Fin Depth a slider between 0-5 ft.

Beyond Defaults: As the user changes Building Area and Number of Floors, Length value is changed by the sqrt(fractionchange), just as you do now. So if Number of Floors goes from 1 to 2, then multiply Length by sqrt(0.5). Each time the Building Area and Number of Floors change, the range of the Length slider changes per the previous equation. The user can modify Length with the slider, and Width is calculated accordingly.

If at any time the Floor Area dimensions change such that 2.2*Perimeter Depth is greater than the dimension they are applied upon, then, as you do now, just don't place interior walls, make it a single zone, and it's up to the user to change dimensions so that interior floors show up again.

**Comment by bwelle** [comment]
@theo-armour 

When you land on the L-Shape page, please put these defaults in this order...Building Area=5000 ft2, Number of Floors=1, Floor Area=calculated, Length uses the equation in the spreadsheet (sqrt(FloorArea*1.8)), Width=sqrt(FloorArea*1.8) (resulting in values of 95 and 95 in this case), Volume=calculated, Floor Height=12ft, Perimeter Depth=10 ft, and Orientation=0deg. As opposed to the box shape, both Length and Width are modifiable and active.

Make Building Area a slider with a range from 1000-100,000 ft2. Increments of 1000 ft. Make Number of Floors a slider from 1-10, increments of 1. Make Length a slider with the ranges calculated in the spreadsheet (min=2/3*sqrt(FloorArea*1.8), max=2*sqrt(FloorArea*1.8) (which this case makes min=63 and max=190). Make Width a slider (increments of 1) with the ranges calculated in the spreadsheet (min=FloorArea/Length), max=10*(FloorArea/Length-(0.9*Length)) (which in this case makes min=52.6 and max=440.8). Make Floor Height a slider between 8-20ft, increments of 1 foot. Make Perimeter Depth a slider between 10-25 feet, increments of 1. Orientation is a slider from 0-350, increments of 10 degrees. All other parameters are calculated fields, no need for a slider.

Then below these parameters have a heading for South, West, East, North. Under each heading put WWR=0.4, # of Windows=4, Window Ratio=0.5, Overhang Depth=3ft, and Fin Depth=0ft.

WWR is a slider between 0-0.9, increments of 0.1. # of Windows a slider between 1-20, increments of 1. Window Ratio a slider between 0.1 and 0.9, increments of 0.1. Overhang Depth a slider between 0-5 ft, Fin Depth a slider between 0-5 ft.

Using the Facade2Windows JS on Spider, call the function (or integrate directly) which takes WWR, # of Windows, and Window Ratio and places an array of windows on that facade. Architects and the Board will love to see this.

Beyond Defaults: As the user changes Building Area and Number of Floors, Length and Width values are changed by the sqrt(fractionchange), just as you do now. So if Number of Floors goes from 1 to 2, then multiply Length by sqrt(0.5). Each time the Building Area and Number of Floors change, the range of the Length and Width slider changes per the previous equations. The user can modify Length with the slider, making the valid range of Width to change and update, and the value of Width should update according to a fractional symmetry between the previous Width range (I give example in spreadsheet.) Then the user can change Width within the valid range, and it changes nothing as it's moved back and forth.

If at any time the Floor Area dimensions change such that 2.2*Perimeter Depth is greater than the dimension they are applied upon, then, as you do now, just don't place interior walls, make it a single zone, and it's up to the user to change dimensions so that interior floors show up again.

For Space naming convention, start with front (and centered if applicable like an H) zone as Zone 1/Space 1. Then go clockwise around shape to number sequentially, with the core zone being last. If 2 floors, then start the next numerical increment at the same spot (i.e. for a L-Shape, 2 floors, Zone 1 will be below Zone 8, etc.)

**Comment by bwelle** [comment]
@theo-armour 

For Adjacent Buildings, remove 2 of the 3 adjacent buildings that are auto-populated. For the landing (default) adjacent building, put offset x=100ft, offset y=100 ft, length=40, width=50, height=80ft, Site orientation to 0deg. Please set Helpers toggle to off as default.

Don't show "The Building" the first time the first the user lands on the page. If they go to shape, then to Adjacent Building, keep "The Building" shown, but don't let the dimensions of it change based on inputs of Adjacent Building page.

**Comment by bwelle** [comment]
@theo-armour 
Additional items

1. I know you set some parameters so when you switch shapes, you retain some values, but I would recommend when you switch from shape to shape, you start with defaults each time (values only don't change when going to Adjacent Building page). I suggest this since keeping some of the values will result in the geometry being invalid under some circumstances.

2. Already noted in Read Me/To Do Generally: Add ortho button for true 2D view. This is critical path.

3. Already noted in Read M/To do Generally: Add XYZ placards once we start having sufficient other inworld text needs. It's critical path to add the XYZ axis in the corner of the grid large enough to be easily seen by the user.

4. Note: Addition of things like being able to click on space number and have it highlighted, and assigning SpaceType to spaces...this functionality recommended by @mdengusiak we do not need for our purposes. I recommend putting this to back burner.

5. Note: We do not need a HUD for this visualization. Only for 3D scatter plot. I suggest putting this to backburner.

6. I had mentioned in prior notes that I wanted the 2D ortho view to show text like "Length, Width, and Thickness", but this is not critical path. So no 2D text needed here now.

7. As mentioned before but not seen in Read Me page, horizontal surfaces (Roof, Floor/Ceiling. BaseSlab must be split into separate surfaces.

8. In previous comments, I wanted to see 3D text for each major orientation (N,W,E,S) shown around the building, so the user would know when they rotate the building, at what point it changes from South for example to West. This is not critical path for the demo, but would be nice to have.

9. gbXML export: Things are looking great here! Once orientation is enabled, the only question is whether or not to put Rectangular Geometry. I have confirmed for Open Studio we do NOT need Rectangular Geometry. However, for our workflow, we DO need the orientation of the surface in question. We can get this from the Rectangular Geometry addition. So from this standpoint, we need it critical path. As I mentioned before, if it's easier just to add a "description" to the surface that puts north, south, east, or west, we can go off that for the demo. Up to you.

I believe as far as I think now, the past 3 posts constitute all the critical path elements for demo. Please let me know if you have questions or would  like to discuss anything. Thanks so much!

**Comment by mdengusiak** [comment]
@bwelle @theo-armour 
In this case when open studio can run without ShellData = rectangular geometry. Can you send me some sample of IDF I see if can use IDF instead of gbXML. I guess, we should call it OSgbXML as this is gbXML version that will only work for OpenStudio so it is not confusing. Please let me know what do you think.

**Comment by bwelle** [comment]
@mdengusiak 
Sure I'll send you IDF from sample file I sent before with rectangular geometry taken out. Adding more to gbXML is not critical path (unless that's the only way to give me surface orientation, which I do need). After the critical path demo, I'm be stoked to have other parts add like shell data so gbXML works for imports into simulation tools other than OS.

**Comment by bwelle** [comment]
@mdengusiak 
[Here](https://perkinswill.box.com/s/2biagyrnu73fy6oyjuo1usjibzrl1ghb) is the IDF generated from OS that ran...

**Comment by bwelle** [comment]
@theo-armour 
How difficult would it be once ortho is enabled to have a view of just the main building with 2D zone labels on it like [this](https://perkinswill.box.com/s/y0ui4vecees9h19wr8q345ah4zz1229g)?

**Comment by mdengusiak** [comment]
@bwelle thanks a lot, I will see if I could use any file type...without shell geometry...

**Comment by theo-armour** [comment]
@bwelle 
> How difficult would it be once ortho is enabled to have a view of just the main building with 2D zone labels on it like this?

I am thinking that gbXML Viewer should be the presentation tool and Build Well the creation tool. I look at the 'Real Time Report' currently in Build Well and feel it's not really a good direction for for a builder. And. in essence, it would mean duplicating much of gbXML viewer into Build Well - which would really slow things down.

Another thought - or just wondering out loud - is how is all this fitting in with the Micro Excel integration? Is the the sort of Flatland graphic that you need the sort of they should be building with their tools?

**Comment by bwelle** [comment]
@theo-armour 
So here' show it's working now...we just finished the initial integration of the Adjacent Building three.js code into the MicroExcel GUI. Basically because they use Angular, they took a sort of hybrid approach by using your three.js code to re-create the logic in JS/html?? if I remember correctly, but still use available functions in three.js from your code when applicable. Thus, somewhat of a hybrid. It will show both 3D interaction and 2D ortho interactive. We imagine basically the same approach for the Geometry shapes, and I have already updated specs to MicroExcel to use the new geometric parameters I provided to you that use the spreadsheet.

We only need the gbXML file export after the user is done with the GUI, spins up their runs, and we run a modified version of your code to take all the inputs and generate a gbXML file that we then use to call the OS API. We then store the gbXML for potential later use. The user in the GUI doesn't want to see all the HUD stuff and properties of surfaces when they hover over it, just raw geometry, then this current set up works fine. Only OS cares about all the jazz in gbXML.

So here's the exception, and the source of my above question...after the Geometry Page in the MicroExcel GUI, we get to a "[Space Layout](https://perkinswill.box.com/s/ecp0qihtm57dikosn82ednke3h161p4v)" page. As you can see, the user must assign space types to each zone, then the total area of each space type is computed. In order to know which space type to apply to what zone (for example they may want Office-Private facing South on Zone 1, they need a visual 2D ortho graphic with the zones labeled. I was hoping we could simply reinstantiate your code here for 2D ortho geometry, but somehow get the text identifiers added to each zone. If you have any innovative thoughts, I'm all ears. The dummy solution is to just put a stock image of the shape with zones numbered, however, if they don't see the relative areas of the zones, it's difficult for them to assign space types to them.

Note: The architect only has Building Area and Program to start with. So they know they want 40% office, 30% retail, and 30% storage. They will want to try to match this. This is just  how they think. So the only way to do that since we auto-perimeter core zone is to see the relative sizes of the zones, the zone number associated with it, then select the space type from the pull down.

**Comment by bwelle** [comment]
@theo-armour 
Maybe have a separate issue for gbXML export? Anyway, I've found this in current version:

Inconsistent view of internal diagonal walls. They seem to disappear then reappear as rotate box.

Diagonal walls not exporting for box shape to gbXML.

Even when dimensions change for a given perimeter depth such that the perimeter depth is no longer valid and no internal walls are generated, gbXML export still exports all the space objects for a floor as if it was perimeter core, not a single zone.
